### PR TITLE
story id #155525079 get Inquiry Count

### PR DESCRIPTION
**Pull request by jitendrajaria**
Inquiries for today
Inquiries for the week
Inquiries for the month
Hot Leads for the month (Inquiries with case index 3 or 4)
Conversions for the month (Inquiries with closing status as enrolled)

### CODE REVIEW

File: src/main/java/creativei/dao/InquiryDao.java [hunk 1]
```diff
@@ -11,10 +13,21 @@
...
-    List<Inquiry> findByInquiryStatusAndIsAttended(InquiryStatus inquiryStatus,Boolean isAttended);
+public interface InquiryDao extends JpaRepository<Inquiry, Long>, JpaSpecificationExecutor<Inquiry> {
+    List<Inquiry> findByInquiryStatus(@Param("status") InquiryStatus status);
+
+    List<Inquiry> findByInquiryStatusAndIsAttended(InquiryStatus inquiryStatus, Boolean isAttended);
+
+    @Query("select count(inq.id) from Inquiry inq where inq.inquiryDate between :inquiryDateTo And :inquiryDateFrom")
+    Integer findCountByInquiryDate(@Param("inquiryDateTo") Date inquiryDateTo, @Param("inquiryDateFrom") Date inquiryDateFrom);
+
+    @Query("select count(inq.id) from Inquiry inq join inq.followUps follUp  where inq.inquiryDate between :inquiryDateTo And :inquiryDateFrom and follUp.caseIndex=:caseIndex3 or follUp.caseIndex=:caseIndex4")
```
Comment (aayushKhandpur) [inline src/main/java/creativei/dao/InquiryDao.java]: Rename param to inquiryToDate and inquiryFromDate

File: src/main/java/creativei/dao/impl/InquiryCustomDaoImpl.java [hunk 2]
```diff
@@ -60,6 +61,20 @@ public Inquiry findOne(Long aLong) {
         return inquiryDao.findOne(aLong);
     }
 
+    @Override
+    public Integer findCountByInquiryDate(Date inquiryDateTo, Date inquiryDateFrom) {
+        return inquiryDao.findCountByInquiryDate(inquiryDateTo,inquiryDateFrom);
+    }
+
+    @Override
```
Comment (aayushKhandpur) [inline src/main/java/creativei/dao/impl/InquiryCustomDaoImpl.java]: What are 2 different case index values?

File: src/main/java/creativei/manager/InquiryManager.java [hunk 3]
```diff
@@ -1,6 +1,7 @@
 package creativei.manager;
 
 
+import com.sun.org.apache.regexp.internal.RE;
```
Comment (aayushKhandpur) [inline src/main/java/creativei/manager/InquiryManager.java]: Do we need this import?

**Review by aayushKhandpur — CHANGES_REQUESTED** [review]
Address reviews

File: src/main/java/creativei/manager/impl/InquiryManagerImpl.java [hunk 4]
```diff
@@ -136,9 +134,27 @@ public ResponseObject getUnattendedInquiry(String boolParam,String statusParam)
         List<InquiryVo> inquiryVos=ResponseHelper.getUnattendedInquiryResponse(inquiryService.getUnattendedInquiry(boolParam,statusParam));
         return ResponseObject.getResponse(inquiryVos);
     }
+
+    @Override
+    public ResponseObject getInquiryCount() {
```
Comment (aayushKhandpur) [inline src/main/java/creativei/manager/impl/InquiryManagerImpl.java:139]: Too many database requests for this request. Optimize queries to reduce db requests

**Review by aayushKhandpur — CHANGES_REQUESTED** [review]
Address review

File: src/main/java/creativei/dao/InquiryCustomDao.java [hunk 5]
```diff
@@ -20,4 +22,10 @@
     List<Inquiry> findAll();
 
     Inquiry findOne(Long aLong);
+
```
Comment (aayushKhandpur) [inline src/main/java/creativei/dao/InquiryCustomDao.java:21]: Also, no need to add a different method for each request. Club monthly and daily request data respectively

Comment (jitendrajaria) [reply src/main/java/creativei/dao/InquiryCustomDao.java:21]: monthly weekly and daily request are clubbed together the other two is for count for enrolled in a month and count of inquiry which have case index 3 or case index 4

File: src/main/java/creativei/dao/impl/InquiryCustomDaoImpl.java (hunk 2 above)
Comment (jitendrajaria) [reply src/main/java/creativei/dao/impl/InquiryCustomDaoImpl.java]: caseIndex 3 "Likely" and caseIndex 4 "HotLead"

File: src/main/java/creativei/dao/InquiryCustomDao.java (hunk 5 above)
Comment (aayushKhandpur) [reply src/main/java/creativei/dao/InquiryCustomDao.java:21]: Case index should be passed as an array of all the values and then fetch for all of them

Comment (jitendrajaria) [reply src/main/java/creativei/dao/InquiryCustomDao.java:21]: ok. The last two can not be clubbed together is that okay?

Comment (aayushKhandpur) [reply src/main/java/creativei/dao/InquiryCustomDao.java:21]: The other 2 can also be subsets of monthly data? If not can be separate requests

Comment (jitendrajaria) [reply src/main/java/creativei/dao/InquiryCustomDao.java:21]: The three monthly,weekly and daily request have same query on database layer. Service layer have different method for each request.

**Comment by aayushKhandpur** [comment]
So are you using same query thrice for the data?

File: src/main/java/creativei/dao/InquiryDao.java (hunk 1 above)
Comment (jitendrajaria) [reply src/main/java/creativei/dao/InquiryDao.java]: name changed e224e82

File: src/main/java/creativei/dao/InquiryCustomDao.java (hunk 5 above)
Comment (jitendrajaria) [reply src/main/java/creativei/dao/InquiryCustomDao.java:21]: case index passed as array 6fa042b

**Review by aayushKhandpur — CHANGES_REQUESTED** [review]
Address reviews

File: src/main/java/creativei/enums/ExceptionType.java [hunk 6]
```diff
@@ -8,7 +8,8 @@
     DUPLICATE_VALUE(1002, "Value already exists"),
     DATABASE_EXCEPTION(1003, "Exception occured while adding values to database"),
     INVALID_METHOD_PARAM(1004,"Invalid Method Request "),
-    DATA_NOT_AVAILABLE(1005,"No record found for requested parameter");
+    DATA_NOT_AVAILABLE(1005,"No record found for requested parameter"),
+    RESULTSET_GENERATION_EXCEPTION(1006,"ResultSet Can not be Generated");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/enums/ExceptionType.java]: This doesn't seem to be a good exception message. Share the reason of failure in user readable text and log the actual exception.

File: src/main/java/creativei/service/impl/InquiryServiceImpl.java [hunk 7]
```diff
@@ -111,4 +120,66 @@ public Inquiry updateAll(List<Inquiry> inquiries) {
...
+            return ResponseObject.getResponse(inquiryCountVo);
+        }catch (Exception e){
+            logger.error("Cause: "+e.getCause().getCause().getMessage());
+            return ResponseObject.getResponse(ExceptionType.RESULTSET_GENERATION_EXCEPTION.getMessage(),ExceptionType.RESULTSET_GENERATION_EXCEPTION.getCode());
+        }
+    }
+
+
+    private Integer getDailyCount() throws ParseException {
+        logger.info("getDailyCount method");
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/impl/InquiryServiceImpl.java]: We are converting the date to string and the converting the same string to date. No need of string.

File: src/main/java/creativei/service/impl/InquiryServiceImpl.java [hunk 8]
```diff
@@ -111,4 +120,66 @@ public Inquiry updateAll(List<Inquiry> inquiries) {
...
+        Date inquiryToDate=LocalizationUtil.stringToDateConverter(date1);
+        Date inquiryFromDate=new Date();
+        return inquiryCustomDao.findCountByInquiryDate(inquiryToDate,inquiryFromDate);
+    }
+
+    private Integer getWeeklyCount() throws ParseException {
+        logger.info("getWeeklyCount method");
+        Date date=new Date();
+        Date inquiryToDate=LocalizationUtil.getWeekStartingDate(date);
+        Date inquiryFromDate=date;
```
Comment (aayushKhandpur) [inline src/main/java/creativei/service/impl/InquiryServiceImpl.java]: Use calendar class object as all these are already provided there.

File: src/main/java/util/LocalizationUtil.java [hunk 9]
```diff
@@ -34,4 +37,24 @@ public static final Date stringtoDateWithTimeConverter(String string) throws Par
             return null;
         return dateFormat.parse(string);
     }
+
+
+    public static final Date getWeekStartingDate(Date date) throws ParseException {
+        SimpleDateFormat dateFormat = new SimpleDateFormat(DATE);
+        Calendar calendar=Calendar.getInstance();
```
Comment (aayushKhandpur) [inline src/main/java/util/LocalizationUtil.java]: Make sure to set time as 00 to cover the whole date.

File: src/main/java/creativei/service/impl/InquiryServiceImpl.java (hunk 7 above)
Comment (jitendrajaria) [reply src/main/java/creativei/service/impl/InquiryServiceImpl.java]: new Date gives the date with current time i have to make sure that date should start with time 00:00 this is the reason why i use string.

File: src/main/java/util/LocalizationUtil.java (hunk 9 above)
Comment (jitendrajaria) [reply src/main/java/util/LocalizationUtil.java]: ya the time is 00 in all starting date

File: src/main/java/creativei/service/impl/InquiryServiceImpl.java (hunk 7 above)
Comment (aayushKhandpur) [reply src/main/java/creativei/service/impl/InquiryServiceImpl.java]: Use calendar class setters. Using that you can set time to 0

Comment (jitendrajaria) [reply src/main/java/creativei/service/impl/InquiryServiceImpl.java]: issue resolved in commit 2df898a

File: src/main/java/creativei/service/impl/InquiryServiceImpl.java (hunk 8 above)
Comment (jitendrajaria) [reply src/main/java/creativei/service/impl/InquiryServiceImpl.java]: Calendar object being used in LocalisationUtil class

**Pull request merged**